File: Buoi_3/2.5.4.undo_app.cpp
#include <iostream>
#include <stack>
#include <string>

class TextEditor {
private:
    std::string text;
    std::stack<std::string> undo_stack;
    std::stack<std::string> redo_stack;

public:
    TextEditor() : text("") {}

    void type_text(const std::string& new_text) {
        undo_stack.push(text);
        text += new_text;
        while (!redo_stack.empty()) redo_stack.pop(); // Clear redo stack
        std::cout << "Đã gõ: " << new_text << std::endl;
        display();
    }

    void undo() {
        if (!undo_stack.empty()) {
            redo_stack.push(text);
            text = undo_stack.top();
            undo_stack.pop();
            std::cout << "Đã hoàn tác hành động." << std::endl;
            display();
        } else {
            std::cout << "Không có hành động nào để hoàn tác." << std::endl;
        }
    }

    void redo() {
        if (!redo_stack.empty()) {
            undo_stack.push(text);
            text = redo_stack.top();
            redo_stack.pop();
            std::cout << "Đã làm lại hành động." << std::endl;
            display();
        } else {
            std::cout << "Không có hành động nào để làm lại." << std::endl;
        }
    }

    void display() {
        std::cout << "Văn bản hiện tại: '" << text << "'\n" << std::endl;
    }
};

int main() {
    TextEditor editor;
    editor.type_text("Hello");
    editor.type_text(" World");
    editor.undo();
    editor.undo();
    editor.redo();
    return 0;
}

File: Buoi_3/2.5.4.undo_app.jl
module TextEditor

mutable struct Editor
    text::String
    undo_stack::Vector{String}
    redo_stack::Vector{String}
end

function Editor()
    return Editor("", String[], String[])
end

function type_text(editor::Editor, new_text::String)
    push!(editor.undo_stack, editor.text)
    editor.text *= new_text
    empty!(editor.redo_stack)  # Clear redo stack
    println("Đã gõ: $new_text")
    display(editor)
end

function undo(editor::Editor)
    if !isempty(editor.undo_stack)
        push!(editor.redo_stack, editor.text)
        editor.text = pop!(editor.undo_stack)
        println("Đã hoàn tác hành động.")
        display(editor)
    else
        println("Không có hành động nào để hoàn tác.")
    end
end

function redo(editor::Editor)
    if !isempty(editor.redo_stack)
        push!(editor.undo_stack, editor.text)
        editor.text = pop!(editor.redo_stack)
        println("Đã làm lại hành động.")
        display(editor)
    else
        println("Không có hành động nào để làm lại.")
    end
end

function display(editor::Editor)
    println("Văn bản hiện tại: '$(editor.text)'\n")
end

end  # module TextEditor

# Minh họa sử dụng
using .TextEditor

# Tạo đối tượng editor và thực hiện các hành động
editor = TextEditor.Editor()
TextEditor.type_text(editor, "Hello")
TextEditor.type_text(editor, " World")
TextEditor.undo(editor)
TextEditor.undo(editor)
TextEditor.redo(editor)
